File: VB/WebSample.Module.Web/CustomizeNavBarController .vb
Imports Microsoft.VisualBasic
Imports System

Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Actions
Imports DevExpress.Persistent.Base
Imports DevExpress.ExpressApp.Web.Templates.ActionContainers
Imports DevExpress.Web.ASPxNavBar
Imports DevExpress.ExpressApp.SystemModule
Imports System.Web.UI.WebControls

Namespace WebSample.Module.Web
	Partial Public Class NavBarViewChangingController
		Inherits WindowController
		Public Sub New()
			InitializeComponent()
			RegisterActions(components)
		End Sub
		Protected Overrides Sub OnActivated()
			MyBase.OnActivated()
			AddHandler Window.TemplateChanged, AddressOf Window_TemplateChanged
		End Sub
		Private Sub Window_TemplateChanged(ByVal sender As Object, ByVal e As EventArgs)
			Dim navBarControl As NavigationBarActionContainer = FindNavBarControl()
			If navBarControl IsNot Nothing Then
				ConfigureNavBar(navBarControl)
			End If
		End Sub
		Private Function FindNavBarControl() As NavigationBarActionContainer
			If Window.Template IsNot Nothing AndAlso Window.IsMain Then
				For Each actionContainer As Object In Window.Template.GetContainers()
					If TypeOf actionContainer Is NavigationBarActionContainer Then
						Return CType(actionContainer, NavigationBarActionContainer)
					End If
				Next actionContainer
			End If
			Return Nothing
		End Function
		Private nodes() As TreeNode
		Private Sub ConfigureNavBar(ByVal navBarControl As NavigationBarActionContainer)
			navBarControl.BeginUpdate()
			Try
				Dim navGroup As NavBarGroup = navBarControl.Groups.FindByText("Data")
				Dim treeView As New MyTreeView(singleChoiceAction1.Items, singleChoiceAction1, nodes)
				nodes = New TreeNode(treeView.Nodes.Count - 1){}
				treeView.Nodes.CopyTo(nodes, 0)
				navGroup.ContentTemplate = treeView
			Finally
				navBarControl.EndUpdate()
			End Try
		End Sub
		Private Sub singleChoiceAction1_Execute(ByVal sender As Object, ByVal e As DevExpress.ExpressApp.Actions.SingleChoiceActionExecuteEventArgs) Handles singleChoiceAction1.Execute
			If e.SelectedChoiceActionItem.Data IsNot Nothing Then
				Dim objectType As Type = ReflectionHelper.FindType(e.SelectedChoiceActionItem.Data.ToString())
				Dim viewID As String = Application.GetListViewId(objectType)
				Dim collectionSource As CollectionSourceBase = Application.CreateCollectionSource(Application.CreateObjectSpace(), objectType, viewID)
				e.ShowViewParameters.TargetWindow = TargetWindow.Current
				e.ShowViewParameters.Context = TemplateContext.ApplicationWindow
				e.ShowViewParameters.CreatedView = Application.CreateListView(viewID, collectionSource, True)
			End If
		End Sub
	End Class
End Namespace

File: CS/WebSample.Module.Web/CustomizeNavBarController .cs
using System;

using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Actions;
using DevExpress.Persistent.Base;
using DevExpress.ExpressApp.Web.Templates.ActionContainers;
using DevExpress.Web.ASPxNavBar;
using DevExpress.ExpressApp.SystemModule;
using System.Web.UI.WebControls;

namespace WebSample.Module.Web{
    public partial class NavBarViewChangingController : WindowController {
        public NavBarViewChangingController() {
            InitializeComponent();
            RegisterActions(components);
        }
        protected override void OnActivated() {
            base.OnActivated();
            Window.TemplateChanged += new EventHandler(Window_TemplateChanged);
        }
        void Window_TemplateChanged(object sender, EventArgs e) {
            NavigationBarActionContainer navBarControl = FindNavBarControl();
            if (navBarControl != null) {
                ConfigureNavBar(navBarControl);
            }
        }
        private NavigationBarActionContainer FindNavBarControl() {
            if (Window.Template != null && Window.IsMain) {
                foreach (object actionContainer in Window.Template.GetContainers()) {
                    if (actionContainer is NavigationBarActionContainer) {
                        return (NavigationBarActionContainer)actionContainer;
                    }
                }
            }
            return null;
        }
        TreeNode[] nodes;
        private void ConfigureNavBar(NavigationBarActionContainer navBarControl) {
            navBarControl.BeginUpdate();
            try {
                NavBarGroup navGroup = navBarControl.Groups.FindByText("Data");
                MyTreeView treeView = new MyTreeView(singleChoiceAction1.Items, singleChoiceAction1, nodes);
                nodes = new TreeNode[treeView.Nodes.Count];
                treeView.Nodes.CopyTo(nodes, 0);
                navGroup.ContentTemplate = treeView;
            }
            finally {
                navBarControl.EndUpdate();
            }
        }
        void singleChoiceAction1_Execute(object sender, DevExpress.ExpressApp.Actions.SingleChoiceActionExecuteEventArgs e) {
            if (e.SelectedChoiceActionItem.Data != null) {
                Type objectType = ReflectionHelper.FindType(e.SelectedChoiceActionItem.Data.ToString());
                string viewID = Application.GetListViewId(objectType);
                CollectionSourceBase collectionSource = Application.CreateCollectionSource(Application.CreateObjectSpace(), objectType, viewID);
                e.ShowViewParameters.TargetWindow = TargetWindow.Current;
                e.ShowViewParameters.Context = TemplateContext.ApplicationWindow;
                e.ShowViewParameters.CreatedView = Application.CreateListView(viewID, collectionSource, true);
            }
        }
    }
}
